indexing-service/src/test/java/io/druid/indexing/overlord/http/OverlordResourceTest.java

```diff
@@ -0,0 +1,335 @@
+  public void testOverlordResource() throws Exception
+  {
+    // basic task master lifecycle test
+    taskMaster.start();
+    announcementLatch.await();
+    while(!taskMaster.isLeading()){
+      // I believe the control will never reach here and thread will never sleep but just to be on safe side
+      Thread.sleep(10);
+    }
+    Assert.assertEquals(taskMaster.getLeader(), druidNode.getHostAndPort());
+    // Test Overlord resource stuff
+    overlordResource = new OverlordResource(taskMaster, new TaskStorageQueryAdapter(taskStorage), null, null, null);
+    Response response = overlordResource.getLeader();
+    Assert.assertEquals(druidNode.getHostAndPort(), response.getEntity());
+
+    String taskId_0 = "0";
+    NoopTask task_0 = new NoopTask(taskId_0, 0, 0, null, null);
+    response = overlordResource.taskPost(task_0);
+    Assert.assertEquals(200, response.getStatus());
+    Assert.assertEquals(ImmutableMap.of("task", taskId_0), response.getEntity());
+
+    // Duplicate task - should fail
+    response = overlordResource.taskPost(task_0);
+    Assert.assertEquals(400, response.getStatus());
+
+    // Task payload for task_0 should be present in taskStorage
+    response = overlordResource.getTaskPayload(taskId_0);
+    Assert.assertEquals(task_0, ((Map) response.getEntity()).get("payload"));
+
+    // Task not present in taskStorage - should fail
+    response = overlordResource.getTaskPayload("whatever");
+    Assert.assertEquals(404, response.getStatus());
+
+    // Task status of the submitted task should be running
+    response = overlordResource.getTaskStatus(taskId_0);
+    Assert.assertEquals(taskId_0, ((Map) response.getEntity()).get("task"));
+    Assert.assertEquals(
+        TaskStatus.running(taskId_0).getStatusCode(),
+        ((TaskStatus) ((Map) response.getEntity()).get("status")).getStatusCode()
+    );
+
+    // Simulate completion of task_0
+    taskCountDownLatches[Integer.parseInt(taskId_0)].countDown();
+    // Wait for taskQueue to handle success status of task_0
+    while (true) {
+      response = overlordResource.getTaskStatus(taskId_0);
+      Assert.assertEquals(200, response.getStatus());
+      if (TaskStatus.success(taskId_0).getStatusCode()
+                    .equals(((TaskStatus) ((Map) response.getEntity()).get("status")).getStatusCode())) {
+        break;
+      }
+      Thread.sleep(10);
+    }
+
+    // Manually insert task in taskStorage
+    String taskId_1 = "1";
+    NoopTask task_1 = new NoopTask(taskId_1, 0, 0, null, null);
+    taskStorage.insert(task_1, TaskStatus.running(taskId_1));
+
+    response = overlordResource.getWaitingTasks();
+    // 1 task that was manually inserted should be in waiting state
+    Assert.assertEquals(1, (((List) response.getEntity()).size()));
+
+    // Simulate completion of task_1
+    taskCountDownLatches[Integer.parseInt(taskId_1)].countDown();
+    // Wait for taskQueue to handle success status of task_1
+    while (true) {
```

Comment:
looks like this while loop and that at line 218 are exactly the same except for taskId, can you move this to a method and use that in both places instead?